feat(database)!: improve orm experience

innocenzi opened the pull request:
Okay, this pull request is huge. It's the result of me trying to build things with the ORM and stumbling upon more and more little things.

It's probably the largest I ever PR'd feature wise, and hopefully the largest you'll ever have to review @brendt!

## Arguments for a hybrid ORM

So, I know you designed the ORM with raw SQL in mind. I personally think it's very valuable to have the ability to map the result of a raw SQL query to an array of models, especially if relationships are mapped as well. That design is great.

However, I also think there is a lot of value in more abstraction, notably when it comes to building queries dynamically _or_ writing very small queries. Here is an example:

```php
// Quite convenient
->whereToday('release_date');

// Works but way less convenient
$start = DateTime::now()->startOfDay()->format('Y-m-d');
$end = DateTime::now()->endOfDay()->format('Y-m-d');
->where('release_date BETWEEN ? and ?', $start, $end);
```

Now, this is a basic example, but imagine having to implement a page with dozens of filters. You'd have to maintain a `$where` and a `$bindings` variable and conditionally concatenate the conditions as well as make sure the final result does not contain syntax errors. Not a lot of developers would be pleased with that.

I also get you wanting to promote raw SQL as the default to embrace it, to promote developer to work more with SQL and think SQL. Personally, I think an ORM is there to make things more convenient. If I wanted to write raw SQL all the time, I wouldn't use one. And maybe that's what you wanted Tempest's ORM to be; an ORM that's almost not one, just there to help with too tedious stuff like relations. However, I don't think this is realistic and I am positive that it would hurt adoption. Most developers will not want to write boilerplate to get their common queries working, they want a quick and nice way to write a query.

The good thing is, thanks to the fact that the ORM is based on the mapper, having an hybrid approach that supports both raw and abstracted SQL is "easy" (putting that in quote because supporting relationships in inserts and updates was not easy lol).

So the initial design was very good in my opinion, but there's an opportunity to have a hybrid approach that would be a win-win to cater to even more developers.

## Changes

- I improved working with `where`
  - I added support for nested `where` using callbacks
  - I added support for `where*` convenience methods for all operations I could think of
  - `where` expected raw SQL previously, now it's `whereRaw`
  - Same for `orderBy`
- I added inline documentation to almost the whole query builder, and the table creation methods as well
- I added some aliases for column creation
- I added a `foreignKey` method for creating a foreign key
- I added a `toRawSql` method (mainly for debugging)
- I fixed serialization of `true`/`false` on MySQL/SQLite, somehow we didn't have that
- I refactored model queries outside of `IsDatabaseModel`
  - `IsDatabaseModel` still exists but no longer feels like it changes the actual model, it's just methods now
  - You can now use the `model` function to get a query builder with methods like `findOrNew` if you don't want to use the trait (though `load`/`save`/`delete`/`update` are only on the trait)
  - The trait no longer forces having an `id` column (see below)
- I improved working with IDs in the ORM:
  - [Models without IDs are now first-class citizens](https://github.com/tempestphp/tempest-framework/issues/1079)
  - Performing operations that require IDs on models that don't have them throw explicit exceptions
  - The ORM no longer assumes the primary key is `id`
  - I renamed `Id` to `PrimaryKey`, which is used to identify the name of the primary key
- [I improved working with DTOs in the ORM](https://github.com/tempestphp/tempest-framework/issues/1286):
  - DTO serialization now works recursively (that was the whole reason I started working on the database, actually)
  - DTO (de)serialization handles a lot more cases, including working with arrays and top-level arrays
- I also reorganized a lot of database tests, removed the legacy `ORM` test directory, and moved most fixtures inside test files directly so it's a lot easier to just jump into a test and find out what's happening

## Considerations

- I wanted to update `Database#fetch` to always return an `ImmutableArray` instead of an array, which would make `fetchFirst` useless as well. I think this would be great but I'm thinking some people might be put off by this. I'm still in favor of it though, so let me know if I should look into that
- Adding first-party UUID generation, would be cool but we can always add that later with no breaking change (I think), and the PR is already big enough

## New updates (06/08)

- I ended up tackling [relation support during inserts/updates](https://github.com/tempestphp/tempest-framework/issues/1087)
  - Doing this still throw explicit exceptions for models without IDs

innocenzi commented:
I've noticed things I need to fix, so I'll put it back on draft

innocenzi commented:
The CI failure is due to `tempest/cryptography` not being published yet on packagist

brendt commented on packages/database/src/Builder/QueryBuilders/HasWhereQueryBuilderMethods.php:
```diff
@@ -2,53 +2,172 @@
+     *
+     * @return self<TModel>
+     */
+    public function where(string $field, mixed $value, string|WhereOperator $operator = WhereOperator::EQUALS): self
```
What about making both possible in `where`? We can detect whether the expression passed in is a single word, and if that's the case, it should call `whereField` behind the scenes?

brendt commented on packages/database/src/Builder/QueryBuilders/WhereGroupBuilder.php:
```diff
@@ -0,0 +1,196 @@
+ * @template TModel of object
+ * @uses \Tempest\Database\Builder\QueryBuilders\HasConvenientWhereMethods<TModel>
+ */
+final class WhereGroupBuilder
```
Oh I just noticed there already is `WhereGroupStatement`. Can't we add all the methods in this class directly over there and skip the class in between?

brendt commented on packages/database/src/GenericDatabase.php:
```diff
@@ -124,6 +128,14 @@ private function resolveBindings(Query $query): array
 
         foreach ($query->bindings as $key => $value) {
+            // Database handle booleans differently. We might need a database-aware serializer at some point.
+            if (is_bool($value)) {
```
Ah, probably not because it needs the dialect?

innocenzi replied:
Yes :/

I don't like this here but it's because it requires more engineering, we can refactor later

brendt commented on packages/database/src/functions.php:
```diff
@@ -2,24 +2,42 @@
+     * @return ModelQueryBuilder<TModel>
      */
-    function model(string|object $model): ModelInspector
+    function model(string $modelClass): ModelQueryBuilder
```
Ah no I understand there's a difference between `ModelQueryBuilder` and `QueryBuilder`. Do we really need these two classes? I rather have one query builder (the query builders always assumed some kind of `model`, so I'm fine with merging the two)

innocenzi replied:
Implemented! Everything use `query` now. Way better.

innocenzi commented on packages/database/src/Builder/QueryBuilders/HasConvenientWhereMethods.php:
```diff
@@ -0,0 +1,498 @@
+     *
+     * @return array{ sql: string, bindings: array<mixed> }
+     */
+    protected function buildCondition(string $fieldDefinition, WhereOperator $operator, mixed $value): array
```
I agree, this deserves a refactor, but that method is also aware of bindings so it's not that easy

innocenzi commented on packages/database/src/Builder/ModelQueryBuilder.php:
```diff
@@ -0,0 +1,332 @@
+     *
+     * @return TModel
+     */
+    public function resolve(string|int|PrimaryKey $id): object
```
We can remove this, I thought that was a method you liked so I kept it 😅

I still have both `get` and `findById` (I find `get` to be confusing personally, but fortunately I added nice comments)

innocenzi commented on packages/database/src/QueryStatements/HasWhereStatements.php:
```diff
@@ -6,7 +6,9 @@
+    /** @var ImmutableArray<WhereStatement> */
     public ImmutableArray $where {
         get;
+        set;
```
Probably not!

innocenzi commented on packages/database/src/Query.php:
```diff
@@ -68,7 +76,95 @@ public function toSql(): string
+    /**
+     * Returns the SQL statement with bindings. This method may generate syntax errors, it is not recommended to use it other than for debugging.
+     */
+    public function toRawSql(): ImmutableString
```
Done, there's a new dedicated `RawSql` class now

innocenzi commented:
> Can you run me through what changed there because it was a bit overwhelming.

## Tests

- I moved around tests, especially in `Builder`, `DtoSerialization`, and `ModelInspector`
- Most of these tests were at the root of `Database` and were a bit tricky to navigate in
- I moved tests from `ORM` to `Database` as well, merging cases in related files
- I added a lot more cases in general
- And I moved fixtures directly in test files so I didn't have to wonder if I was going to break other tests when updating fixtures for new cases

## `InsertQueryBuilder` and `UpdateQueryBuilder`

Most of the changes have to do with supporting relationships insertion/updating in different scenarios, as well as handling situations where there is no primary key. I also took some time to try and clean up the logic in separate methods because it was becoming complex.

I think the best way to review is to checkout the PR locally and look around these files yourself 😬

innocenzi commented:
Thanks for the thoughts @moebrowne! I think you're trying to find a nice middleground between this PR and Brent's opinion, but I'm convinced the default `where` experience should be the most convenient one. Otherwise, this will increase friction and will be detrimental to Tempest.

I sincerely believe that `whereRaw` is the best option. It's really not that bad to type. It's consistent with expectations (not that this is a criteria for Tempest's features), it has clear naming and the same functionality as before.
